Selected in View within Gallery

rohinish404 opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [X] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission


## Have you discussed this change with the InvokeAI team?
- [X] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [ ] No


## Description


## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below. 

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #4778 
- Closes #4778 

## QA Instructions, Screenshots, Recordings

<!-- 
Please provide steps on how to test changes, any hardware or 
software specifications as well as any other pertinent information. 
-->

## Added/updated tests?

- [ ] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

lstein reviewed (changes requested):
Tested and it works as advertised. However, the behavior is that the selected thumbnail is always on the top row of the gallery. I would personally prefer behavior in which the gallery scrolls up to keep the selected thumbnail in view _only after the selected thumbnail reaches the bottom of the gallery_. I find it useful to see the selected thumbnail in context of what both precedes and follows it.

psychedelicious commented on invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImageGrid.tsx:
```diff
@@ -48,6 +58,10 @@ const GalleryImageGrid = () => {
 
+  const lastSingleSelectionImage = useAppSelector(selector);
+
+  const imgRef = useRef<VirtuosoGridHandle>(null);
```
Suggest a `virtuosoRef` as a clearer name for this

rohinish404 replied:
fixed with latest commit!

psychedelicious commented:
I agree with @lstein, it should only scroll when it needs to. I think you can use the `rangeChanged` callback to accomplish this by comparing the current visible range with the selected item index.

rohinish404 commented:
> Tested and it works as advertised. However, the behavior is that the selected thumbnail is always on the top row of the gallery. I would personally prefer behavior in which the gallery scrolls up to keep the selected thumbnail in view _only after the selected thumbnail reaches the bottom of the gallery_. I find it useful to see the selected thumbnail in context of what both precedes and follows it.

@lstein Fixed with latest commit!

psychedelicious commented:
Marking this as a draft pending resolution of issues as discussed in https://discord.com/channels/1020123559063990373/1177849030911479808

lstein reviewed (approved):
I've tested the recent changes and everything seemed to work:
1. Selecting next and previous thumbnails using right and left arrow keys.
2. Scrolling up and down in the gallery using up/down arrow keys
3. Scrolling with mouse wheel
4. Scrolling current selected thumbnail out of sight and then using right arrow to select next thumbnail scrolls the selected thumbnail into view.

The only oddity I noticed is that when you select a thumbnail it is outlined in blue as expected. However, when using the right arrow to select the next thumbnail, the outline turns to white. This is also present in `main`, so not specific to the current PR.

rohinish404 commented:
> I've tested the recent changes and everything seemed to work:
> 
> 1. Selecting next and previous thumbnails using right and left arrow keys.
> 2. Scrolling up and down in the gallery using up/down arrow keys
> 3. Scrolling with mouse wheel
> 4. Scrolling current selected thumbnail out of sight and then using right arrow to select next thumbnail scrolls the selected thumbnail into view.
> 
> The only oddity I noticed is that when you select a thumbnail it is outlined in blue as expected. However, when using the right arrow to select the next thumbnail, the outline turns to white. This is also present in `main`, so not specific to the current PR.

Thanks for the review!! So no further changes are required?

psychedelicious reviewed (approved):
I updated the branch and made two small improvements:
- In `useNextPrevImage`, do not create new variables for all the refs. We can use them directly to reduce memory usage and improve performance. It's a tiny improvement.
- In `useNextPrevImage`, the the prev / next callbacks only aligned to 'start' and 'end' respectively. This resulted in the alignment being different depending on which arrow you press, but that's not correct. I extracted the logic from `useScrollToVisible` that determines which alignment to use, and also used it in `useNextPrevImage`

I also make two changes related to network requests:
- Set the fallback debounce of metadata fetching hook of 300ms (was 0ms)
- Add a debounced hook for workflow, and set it to 300ms

This *drastically* reduces the number of network requests made when navigating the gallery with arrow keys without impacting user experience.

Thanks for working on this @rohinish404 !

rohinish404 commented:
> I updated the branch and made two small improvements:
> 
> * In `useNextPrevImage`, do not create new variables for all the refs. We can use them directly to reduce memory usage and improve performance. It's a tiny improvement.
> * In `useNextPrevImage`, the the prev / next callbacks only aligned to 'start' and 'end' respectively. This resulted in the alignment being different depending on which arrow you press, but that's not correct. I extracted the logic from `useScrollToVisible` that determines which alignment to use, and also used it in `useNextPrevImage`
> 
> I also make two changes related to network requests:
> 
> * Set the fallback debounce of metadata fetching hook of 300ms (was 0ms)
> * Add a debounced hook for workflow, and set it to 300ms
> 
> This _drastically_ reduces the number of network requests made when navigating the gallery with arrow keys without impacting user experience.
> 
> Thanks for working on this @rohinish404 !

Woah..that's some awesome improvements! Thanks for merge and helping out throughout this issue! @psychedelicious and @lstein
